### ISSUE TITLE
Back camera VisualEffect

### ISSUE THREAD

**Issue by pzs7602** [shogo4405/lf.swift#72]
@shogo4405 :
Front camera VisualEffect is functional while in AVCaptureSession():
```swift
// example Custom VisualEffect
final class CameraMixEffect: VisualEffect, AVCaptureVideoDataOutputSampleBufferDelegate {
    let filter:CIFilter? = CIFilter(name: "CISourceOverCompositing")
    var camera:CIImage?
    var lockQueue:dispatch_queue_t? dispatch_queue_create(
         "CameraMixEffect.lock", DISPATCH_QUEUE_SERIAL
   )

    override init() {
        super.init()
    }

    override func execute(image: CIImage) -> CIImage {
        guard let filter:CIFilter = filter else {
            return image
        }
        dispatch_async(lockQueue) {
            self.filter.setValue(camera, forKey: "inputImage")
            self.filter.setValue(image, forKey: "inputBackgroundImage")
        }
        return filter.outputImage!
    }

    func captureOutput(captureOutput: AVCaptureOutput!, didOutputSampleBuffer sampleBuffer: CMSampleBuffer!, fromConnection connection: AVCaptureConnection!) {
        guard let image:CVImageBufferRef = CMSampleBufferGetImageBuffer(sampleBuffer) else {
            return
        }
        camera = CIImage(CVPixelBuffer:  image)
    }
}
```
```swift
// next 
var session = AVCaptureSession()
var videoOutput =  AVCaptureVideoDataOutput()
session.addInput(try! AVCaptureDeviceInput(device: camera))
session.addOutput(videoDataOutput)

var effect = CameraMixEffect()
videoOutput.setSampleBufferDelegate(effect, queue: effect.lockQueue)

var stream:RTMPStream = RTMPStream()
stream.registerEffect(effect)
```
But is it possible to use back camera ? How ?

**Comment by shogo4405** [comment]
I think back and front camera can not be used at the same time.
I will google it.

**Comment by pzs7602** [comment]
I mean not at the same time

**Comment by shogo4405** [comment]
what's the meaning?
Do you want to publish ScreenCaptureSession with Camera?

**Comment by pzs7602** [comment]
@shogo4405: Do you want to publish ScreenCaptureSession with Camera?
Yes, especially with back camera

**Comment by shogo4405** [comment]
@pzs7602 Sorry. ScreenCaptureSession with VisualEffect is not support yet. I forget it.... Please change here. Then works OK. I will change here in a few days.

### ScreenCaptureSession.swift
```swift
public final class ScreenCaptureSession: NSObject {
    static let defaultFrameInterval:Int = 2
    static let defaultAttributes:[NSString: NSObject] = [
// kCVPixelBufferPixelFormatTypeKey: NSNumber(unsignedInt:kCVPixelFormatType_420YpCbCr8BiPlanarVideoRange),
        kCVPixelBufferPixelFormatTypeKey: NSNumber(unsignedInt: kCVPixelFormatType_32BGRA),
        kCVPixelBufferCGBitmapContextCompatibilityKey: true
    ]
```
### AACEncoder.swift 
```swift
// MARK: ScreenCaptureOutputPixelBufferDelegate
extension VideoIOComponent: ScreenCaptureOutputPixelBufferDelegate {
    func didSetSize(size: CGSize) {
        dispatch_async(lockQueue) {
            self.encoder.width = Int32(size.width)
            self.encoder.height = Int32(size.height)
        }
    }
    func pixelBufferOutput(pixelBuffer:CVPixelBufferRef, timestamp:CMTime) {
        encoder.encodeImageBuffer(
            // pixelBuffer, 
            effects.isEmpty ? pixelBuffer : effect(pixelBuffer),
            // ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
            presentationTimeStamp: timestamp,
            presentationDuration: timestamp
        )
    }
}
```
Test CameraEffect
```swift
// MARK: - CameraEffect
public class CameraEffect: VisualEffect {
    private var session:AVCaptureSession?
    private var capture:CIImage?
    private var dataOutput:AVCaptureVideoDataOutput?
    private let filter:CIFilter? = CIFilter(name: "CISourceOverCompositing")
    private let lockQueue:dispatch_queue_t = dispatch_queue_create("com.github.shogo4405.lf.CameraEffect", nil)

    public func attachCamera(camera:AVCaptureDevice?) {
        dataOutput = AVCaptureVideoDataOutput()
        dataOutput?.setSampleBufferDelegate(self, queue: lockQueue)
        session = AVCaptureSession()
        session?.addInput(try! AVCaptureDeviceInput(device: camera))
        session?.addOutput(dataOutput)
        session?.startRunning()
    }

    public override func execute(image: CIImage) -> CIImage {
        guard let filter:CIFilter = filter else {
            return image
        }

        var result:CIImage?
        dispatch_sync(lockQueue) {
            filter.setValue(self.capture, forKey: "inputImage")
            filter.setValue(image, forKey: "inputBackgroundImage")
            result = filter.outputImage
        }

        return result!
    }
}

// MARK: AVCaptureVideoDataOutputSampleBufferDelegate
extension CameraEffect: AVCaptureVideoDataOutputSampleBufferDelegate {
    public func captureOutput(captureOutput: AVCaptureOutput!, didOutputSampleBuffer sampleBuffer: CMSampleBuffer!, fromConnection connection: AVCaptureConnection!) {
        guard let image:CVImageBufferRef = CMSampleBufferGetImageBuffer(sampleBuffer) else {
            return
        }
        capture = CIImage(CVPixelBuffer: image)
        capture = capture!.imageByApplyingTransform(CGAffineTransformMakeScale(0.1, 0.1))
    }
}
```

**Comment by shogo4405** [comment]
released 0.3.4
